### PR TITLE
Add progress bar in GUI

### LINKED ISSUE #1500: GUI: integrate new progress reporter as progress bars
- In the UDF running engine of the web API, use a custom progress reporter
- Forward events from the progress reporter via the websocket connection
- Display progress bar in the web GUI, in some form or other
    - don't flash progress bar if operation finishes "fast" (under some threshold, let's say 200ms?)
    - ...but do show progress bar even before the first result comes in, which is currently a bit confusing

### PR DESCRIPTION
**Pull request by sk1p**
It's a combined progress bar for all currently running jobs (which also means it may jump backwards if one of the jobs finishes...)

Closes #1500 

(note that #1513 needs to be fixed before rebuilding the client, which will be done in a separate PR)

## Contributor Checklist:

* [ ] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [ ] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

## Reviewer Checklist:

* [x] `/azp run libertem.libertem-data` passed
* [ ] No import of GPL code from MIT code

### CODE REVIEW

**Comment by matbryan52** [comment]
Great! I had a look over the code (for as much as I can follow the TS), and seems like it wasn't too hard to integrate with the `ProgressReporter`. Looking forward to seeing it running once the support PRs are merged.

**Comment by sk1p** [comment]
> Great! I had a look over the code (for as much as I can follow the TS), and seems like it wasn't too hard to integrate with the `ProgressReporter`. Looking forward to seeing it running once the support PRs are merged.

Thanks for having a look! I enjoyed writing TypeScript after a while, the type system is really quite enjoyable.

I think the main gotcha was that the callbacks in the `ProgressReporter` are not running on the main thread, so we have to do some explicit gymnastics to dispatch the websocket message on the right thread and message loop. Once that was solved it was only a matter of state mgmt and viewing progress in the GUI.

**Comment by sk1p** [comment]
> Looking forward to seeing it running once the support PRs are merged.

It's now rebased on top of the changes merged with #1513, and I've pushed a rebuilt client. Should now be testable without running vite.

File: client/src/progress/components/ProgressPie.tsx
Comment (matbryan52) [inline client/src/progress/components/ProgressPie.tsx:1]: I can't figure out if `ProgressPie` is used anywhere? If not then maybe we can remove it for now?

**Review by matbryan52 — APPROVED** [review]
I like it ! I can't really comment on the TypeScript, but it certainly seems consistent with my understanding of the rest of the code.

While testing I did notice that I think we have a problem with the `needs_rerun` mechanism - it seems to work for the Disk analysis but Ring and Point, for example, always re-calculate (for me). This is happening on current master, too, so I think it's not related to the progress bar.

File: client/src/progress/components/ProgressPie.tsx
Comment (sk1p) [reply client/src/progress/components/ProgressPie.tsx:1]: Good catch, that was from my first go at it - but only having the Pie is too subtle. I'll clean this up.

**Comment by sk1p** [comment]
Thanks for having a look!

> I like it ! I can't really comment on the TypeScript, but it certainly seems consistent with my understanding of the rest of the code.

Yeah, review regarding the functionality is more important than reviewing the Typescript impl in this case ;)

> While testing I did notice that I think we have a problem with the `needs_rerun` mechanism - it seems to work for the Disk analysis but Ring and Point, for example, always re-calculate (for me). This is happening on current master, too, so I think it's not related to the progress bar.

Interesting - I'll have a look if I can reproduce this - maybe there's something inconsistent in how parameters are sent over or compared? I don't remember how I implemented this, so I'll need to have a look at the code.

**Pull request merged**